### ISSUE TITLE
[tune][rllib] Hyperparameter sweeps not running asynchronously

### ISSUE THREAD

**Issue by Andrej4156** [ray-project/ray#12073]
### What is the problem?
Running several workers in parallel on separate environments results in each worker running at a fraction of the speed a single worker would run at. This problem occurs in both my custom environment and the cartpole environment though to a lesser extent in the latter.

For example, suppose I run a single process with 1 worker. It evaluates 600 timesteps every second. If I add a second it evaluates 900 timesteps every second. We can continue this up to 6 workers with diminishing returns and end up evaluating 1,200 timesteps every second. So far so good, this is all to be expected.

Suppose now I want to run a sweep for learning rate and have those 6 workers run independently. Each runs a separate experiment and should be entirely separate from the other 5 workers so one would logically expect that each would run at around 600 timesteps. However, this is not the case. Instead each runs at 1/6 the combined speed and so only evaluates 200 timesteps every second. This means that there is literally no advantage to running these operations in parallel. Running the same 6 experiments with all 6 workers working together sequentially would take the same amount of time.

Testing this in the Cartpole environment shows the following:

1 worker running on one environment
![image](https://user-images.githubusercontent.com/64599107/99415314-ac574180-28c5-11eb-8a01-ce21cc3c4622.png)
Total timesteps/sec: 963
timestep/sec per worker: 963

6 workers running on one environment
![image](https://user-images.githubusercontent.com/64599107/99415470-d27ce180-28c5-11eb-9cb3-6fab840a97a0.png)
Total timesteps/sec: 2150
timestep/sec per worker: 358

6 workers each running their own experiment
![image](https://user-images.githubusercontent.com/64599107/99415703-1839aa00-28c6-11eb-91de-fa6a855dfe4a.png)
Total timesteps/sec: 3017
timestep/sec per worker: 503

So it's not as bad as it is in my custom environment but the parallelization gains seem awful. It would seem that I would see significantly better results loading up 6 separate docker containers than using Ray's built in tools. Note: I have tested this on 3 separate computers so I know this isn't a hardware issue.

*Ray version 0.8.7, Python 3.6.9, TensorFlow 2.3.0, Ubuntu 18.04.4 LTS

### Reproduction (REQUIRED)

    import ray
    from ray.tune.schedulers import AsyncHyperBandScheduler
    import random
    import numpy as np
    from ray import tune
    from ray.tune import sample_from
    from ray.rllib.contrib.alpha_zero.environments.cartpole import CartPole

    def main():
        ray_init = {
            'local_mode': False,
            'include_dashboard': True,
            'dashboard_host': '0.0.0.0',
        }

    async_hb_scheduler = AsyncHyperBandScheduler(
        time_attr='training_iteration',
        metric='reward_mean',
        mode='max',
        max_t=10,
        grace_period=2,
        reduction_factor=4,
        brackets=1)
    tune_run = {

        'run_or_experiment': 'PPO',
        'name': 'glider1',
        'local_dir': '/opt/data/ray_results/',
        'checkpoint_freq': 10,
        'checkpoint_at_end': True,
        'num_samples': 6,
        'scheduler': async_hb_scheduler,

        'config': {
            'env_config': {
                'episode_num': 0,
            },
            "num_cpus_for_driver": 1,
            "num_gpus": 0,
            "num_workers": 1,
            "num_cpus_per_worker": 1,
            "num_gpus_per_worker": 0,
            "env": CartPole,
            "lr": sample_from(lambda _: 10 ** np.random.uniform(-4, -6)),
            "vf_loss_coeff": 100,
        },
    }
    ray.init(**ray_init)
    analysis = tune.run(**tune_run)
    print(analysis)

    if __name__ == "__main__":
        main()

```

- [ x ] I have verified my script runs in a clean environment and reproduces the issue.

**Comment by krfricke** [comment]
Hi Andrej, thanks for raising this issue. Just to clarify, are you running this on a single machine or on a cluster (it looks like a single machine)? How many CPUs are on your machine? 

Parallelization can only do so much on a single machine - if your CPUs are at 100% utilization, it doesn't matter if they're working on six workers with one environment each, or one worker with six environments. There might be some differences due to data flow, but overall throughput will be similar.

**Comment by Andrej4156** [comment]
The above example is running on a single machine with an Intel 8700K processor (6 cores, 12 threads) but this also occurs on a server with 40 cores, 80 threads.

Generally the CPUs are below 100% utilization. Even when running a single environment on the single machine with 10 workers, the average CPU utilization is well below 50%. One core might spike to 100% and they might all jump up to 80% for a second but they generally spend most of their time around 30%.

To be clear though, the issue is not that I want 6 different workers to run 6 different environments on the same experiment (where they would have to share data). The issue is that I want to run 6 different experiments simultaneously where workers do not need to share information. There should be no CPU bottleneck, especially on the server.

**Comment by ericl** [comment]
@richardliaw this is probably related to the event loop bottleneck. It might be fixed by your latest patch.

**Comment by richardliaw** [comment]
@Andrej4156 Can you try again with the latest nightly wheels? `pip install -U [wheel-link]` https://docs.ray.io/en/master/installation.html#daily-releases-nightlies

**Comment by Andrej4156** [comment]
I rebuilt my docker container with the nightly version. I had to copy over your requirements.txt to get it to work. I've posted it below because while I think I did it right I've never actually built a docker container with a wheel link.

[requirements.txt](https://github.com/ray-project/ray/files/5668272/requirements.txt)

In terms of performance, I reran all three tests in the original post and saw no improvement in cartpole but potentially saw some small improvement in my own custom environment. Using 6 workers on 6 environments evaluated roughly 20% more timesteps than 6 workers on one environment but this isn't something I specifically tested last time (this time ran without the GPU). It's not much but it's something.

Unfortunately there's an added wrinkle in all this. I have an Intel i7-8700K (6 core, 12 threads) processor so I can only run up to 6 workers in parallel. I can however assign 10 workers to one environment and enable the GPU (GTX 1070TI)  which allows me to match those parallelized performance gains. Enabling the GPU on the 6 workers 6 environments model cuts performance in half. I'm not sure if this is a graphics card bottleneck or something else. To be clear, by enabling the GPU I mean setting ```"num_gpus"``` in the config to 1. ```"num_gps_per_worker``` stays at 0.

To summarize the results on my environment:
- Running 6 workers on 6 environments without GPU: ~1050 it/s 
- Running 6 workers on 1 environment without GPU: ~850 it/s
- Running 10 workers on 1 environment without GPU: ~850 it/s
- Running 10 workers on 1 environment with GPU: ~1050 it/s 


Interestingly though, I noticed this line coming up over and over in the output this time around.
```2020-12-09 19:48:37,867	WARNING util.py:142 -- The `process_events` operation took 4.564 s, which may be a performance bottleneck.```

I'm not sure if it was there before but I noticed it this time around in both my custom environment and cartpole. Cartpole's ```process_events``` operation generally took 2-2.5 seconds. My custom environment was generally in the 4-5 second range. When I ran multiple samples in parallel this number varied from as low as 0.6 seconds to as high as 11.8 seconds.



On another note, I was having a separate issue with Hyperband never cutting off trials which seems to be fixed now so thank you for fixing that.

**Comment by ericl** [comment]
> I'm not sure if it was there before but I noticed it this time around in both my custom environment and cartpole. Cartpole's process_events operation generally took 2-2.5 seconds. My custom environment was generally in the 4-5 second range. When I ran multiple samples in parallel this number varied from as low as 0.6 seconds to as high as 11.8 seconds.

That seems high, likely a bug in Tune, perhaps it's writing to a slow NFS filesystem or something to update result logs.

To mitigate it, you can increase the length of your train call. As a rule of thumb for large scale runs I always set `min_iter_time_s` to at least 60, which means Tune only has to process one event per minute per trial.

**Comment by Andrej4156** [comment]
Setting ```"min_iter_time_s"``` to 60 results in the operation taking significantly longer

```2020-12-09 20:45:14,730	WARNING util.py:142 -- The `process_events` operation took 62.502 s, which may be a performance bottleneck.```

**Comment by ericl** [comment]
Ah looking at the code, that warning is incorrect. @richardliaw can we remove it? It's covering a blocking call to wait for the next trial result.

**Comment by Andrej4156** [comment]
Do you have any other suggestions for increasing the training speed? I'm still confused why Cartpole has so much better parallelization than my my custom environment. I know it must be something to do with my environment but I don't know what.

**Comment by krfricke** [comment]
cc @sven1977

**Comment by sven1977** [comment]
Would you be able to share your custom environment @Andrej4156 ?

**Comment by richardliaw** [comment]
Closing this for now as we should have much better scalability now in 1.3+
